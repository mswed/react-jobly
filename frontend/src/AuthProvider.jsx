import React, { useState, useEffect } from 'react';
import JoblyApi from './api';

// We create the context
const AuthContext = React.createContext();

// We set up a component to manage the context
const AuthProvider = ({ children }) => {
  // We initialize the token with the value from localStorage
  const [token, setToken] = useState(localStorage.getItem('jobly-token'));

  // Whenever the token changes (login/logout) we need to update
  // both the api module and localStorage
  useEffect(() => {
    JoblyApi.token = token;
    localStorage.setItem('jobly-token', token || '');
  }, [token]);

  const login = async (username, password) => {
    try {
      const token = await JoblyApi.login(username, password);
      setToken(token);
      return { success: true };
    } catch (error) {
      return { success: false, error };
    }
  };

  const register = async (username, password, firstName, lastName, email) => {
    try {
      const token = await JoblyApi.register(username, password, firstName, lastName, email);
      setToken(token);
      return { success: true };
    } catch (error) {
      return { success: false, error };
    }
  };

  // Create the value object
  const value = { token, register };

  // Wrap everything in the context
  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};

export { AuthContext, AuthProvider };
