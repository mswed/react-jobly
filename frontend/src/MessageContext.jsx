import React, { useState } from 'react';

// We create the context
const MessageContext = React.createContext();

// We set up a component to manage the context
const MessagesProvider = ({ children }) => {
  const [message, setMessage] = useState(null);

  const showMessage = (text, variant = 'success', duration = 5000) => {
    setMessage({ text, variant, duration });
    // Auto clear
    if (duration) {
      setTimeout(() => {
        clearMessage();
      }, duration);
    }
  };

  const clearMessage = () => {
    setMessage(null);
  };

  return <MessageContext.Provider value={{ message, showMessage, clearMessage }}>{children}</MessageContext.Provider>;
};

export { MessageContext, MessagesProvider };
