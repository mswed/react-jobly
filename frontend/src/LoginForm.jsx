import { Container, Form, Button, Row, Col, Card } from 'react-bootstrap';
import { useState, useContext } from 'react';
import { AuthContext } from './AuthProvider';
import { useNavigate } from 'react-router-dom';

const LoginForm = () => {
  const navigate = useNavigate();
  const INITIAL_STATE = {
    username: '',
    password: '',
  };

  // Get functions from AuthContext
  const { login } = useContext(AuthContext);

  // Manage the form inputs
  const [loginForm, setLoginForm] = useState(INITIAL_STATE);

  const handleChange = (evt) => {
    const { name, value } = evt.target;
    setLoginForm((originalData) => ({ ...originalData, [name]: value }));
  };

  const handleSubmit = async (evt) => {
    evt.preventDefault();
    const success = await login(loginForm.username, loginForm.password);
    if (success) {
      navigate('/');
    }
  };

  return (
    <Form onSubmit={handleSubmit}>
      <Form.Group className="mb-3" controlId="username">
        <Form.Label>Username</Form.Label>
        <Form.Control placeholder="Username" name="username" onChange={handleChange} value={loginForm.username} />
      </Form.Group>
      <Form.Group className="mb-3" controlId="password">
        <Form.Label>Password</Form.Label>
        <Form.Control placeholder="Password" name="password" onChange={handleChange} value={loginForm.password} />
      </Form.Group>
      <Button type="submit" className="w-100">
        Login
      </Button>
    </Form>
  );
};

export default LoginForm;
