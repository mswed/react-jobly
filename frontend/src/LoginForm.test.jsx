import { it, expect, vi, describe, beforeEach } from 'vitest';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import LoginForm from './LoginForm';
import { MemoryRouter } from 'react-router-dom';
import { MessageContext } from './MessageContext';
import { AuthContext } from './AuthProvider';

// Mock hook from react-router-dom. First we create a mock version of useNavigate
const mockNavigate = vi.fn();
// Then tell vite to intercept the import of react-router-dom and replace the hook with
// our mocked version
vi.mock('react-router-dom', async () => {
  const actual = await vi.importActual('react-router-dom');
  return {
    ...actual,
    useNavigate: () => mockNavigate,
  };
});

describe('LoginForm', () => {
  // Reset all mocks before each test
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders without crashing', () => {
    render(
      <MemoryRouter>
        <MessageContext.Provider value={{ showMessage: vi.fn() }}>
          <AuthContext.Provider value={{ login: vi.fn() }}>
            <LoginForm />
          </AuthContext.Provider>
        </MessageContext.Provider>
      </MemoryRouter>
    );
  });

  it('has username and password inputs and a submit button', () => {
    render(
      <MemoryRouter>
        <MessageContext.Provider value={{ showMessage: vi.fn() }}>
          <AuthContext.Provider value={{ login: vi.fn() }}>
            <LoginForm />
          </AuthContext.Provider>
        </MessageContext.Provider>
      </MemoryRouter>
    );

    // Check for form elements
    expect(screen.getByLabelText(/Username/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/Password/i)).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /Login/i })).toBeInTheDocument();
  });

  it('updates form state when typing', () => {
    render(
      <MemoryRouter>
        <MessageContext.Provider value={{ showMessage: vi.fn() }}>
          <AuthContext.Provider value={{ login: vi.fn() }}>
            <LoginForm />
          </AuthContext.Provider>
        </MessageContext.Provider>
      </MemoryRouter>
    );

    // Get form inputs
    const usernameInput = screen.getByLabelText(/Username/i);
    const passwordInput = screen.getByLabelText(/Password/i);

    // Type in inputs
    fireEvent.change(usernameInput, { target: { value: 'testuser' } });
    fireEvent.change(passwordInput, { target: { value: 'password123' } });

    // Check values were updated
    expect(usernameInput).toHaveValue('testuser');
    expect(passwordInput).toHaveValue('password123');
  });

  it('calls login with form data on submit', async () => {
    // Mock functions
    const mockLogin = vi.fn().mockResolvedValue({ success: true });
    const mockShowMessage = vi.fn();

    render(
      <MemoryRouter>
        <MessageContext.Provider value={{ showMessage: mockShowMessage }}>
          <AuthContext.Provider value={{ login: mockLogin }}>
            <LoginForm />
          </AuthContext.Provider>
        </MessageContext.Provider>
      </MemoryRouter>
    );

    // Fill form
    fireEvent.change(screen.getByLabelText(/Username/i), {
      target: { value: 'testuser' },
    });
    fireEvent.change(screen.getByLabelText(/Password/i), {
      target: { value: 'password123' },
    });

    // Submit form
    fireEvent.click(screen.getByRole('button', { name: /Login/i }));

    // Check login was called with correct data
    await waitFor(() => {
      expect(mockLogin).toHaveBeenCalledWith('testuser', 'password123');
    });
  });

  it('navigates and shows success message on successful login', async () => {
    // Mock functions for successful login
    const mockLogin = vi.fn().mockResolvedValue({ success: true });
    const mockShowMessage = vi.fn();

    render(
      <MemoryRouter>
        <MessageContext.Provider value={{ showMessage: mockShowMessage }}>
          <AuthContext.Provider value={{ login: mockLogin }}>
            <LoginForm />
          </AuthContext.Provider>
        </MessageContext.Provider>
      </MemoryRouter>
    );

    // Fill and submit form
    fireEvent.change(screen.getByLabelText(/Username/i), {
      target: { value: 'testuser' },
    });
    fireEvent.change(screen.getByLabelText(/Password/i), {
      target: { value: 'password123' },
    });
    fireEvent.click(screen.getByRole('button', { name: /Login/i }));

    // Check results
    await waitFor(() => {
      expect(mockNavigate).toHaveBeenCalledWith('/');
      expect(mockShowMessage).toHaveBeenCalledWith('Login Successfull', 'success');
    });
  });

  it('shows error message on failed login', async () => {
    // Mock functions for failed login
    const mockLogin = vi.fn().mockResolvedValue({ success: false });
    const mockShowMessage = vi.fn();

    render(
      <MemoryRouter>
        <MessageContext.Provider value={{ showMessage: mockShowMessage }}>
          <AuthContext.Provider value={{ login: mockLogin }}>
            <LoginForm />
          </AuthContext.Provider>
        </MessageContext.Provider>
      </MemoryRouter>
    );

    // Fill and submit form
    fireEvent.change(screen.getByLabelText(/Username/i), {
      target: { value: 'baduser' },
    });
    fireEvent.change(screen.getByLabelText(/Password/i), {
      target: { value: 'wrongpass' },
    });
    fireEvent.click(screen.getByRole('button', { name: /Login/i }));

    // Check error handling
    await waitFor(() => {
      expect(mockNavigate).toHaveBeenCalledWith('/');
      expect(mockShowMessage).toHaveBeenCalledWith('Login Failed! Incorrect username or password!', 'danger');
    });
  });
});
