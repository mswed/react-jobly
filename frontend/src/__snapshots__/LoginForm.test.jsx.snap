// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="justify-content-center mt-5 row"
    >
      <div
        class="col-lg-6 col-md-8"
      >
        <h3
          class="text-shadow"
        >
          Login
        </h3>
        <div
          class="shadow-lg  p-3 card"
        >
          <form
            class=""
          >
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="username"
              >
                Username
              </label>
              <input
                class="form-control"
                id="username"
                name="username"
                placeholder="Username"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="password"
              >
                Password
              </label>
              <input
                class="form-control"
                id="password"
                name="password"
                placeholder="Password"
                type="password"
                value=""
              />
            </div>
            <button
              class="w-100 btn btn-primary"
              type="submit"
            >
              Login
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
